AzureIoTHub.agent.lib.nut

```diff
@@ -629,513 +632,819 @@ class AzureIoTHub {
+                local sas = AzureIoTHub.SharedAccessSignature.create(
+                    resourceUri, null, _connStrParsed.SharedAccessKey, passwDeadTime).toString();
 
-        function _openAuthSession() {
-            _sessions.auth <- _connection.opensession(_amqpAuthSessionStatusHandler.bindenv(this));
-        }
+                local options = {
+                    "username" : username,
+                    "password" : sas
+                };
 
-        function _openAuthReceiver() {
-            _receivers.auth <- _sessions.auth.openreceiver(AZURE_CLAIM_BASED_SECURITY_PATH, _amqpAuthReceiverStatusHandler.bindenv(this), _authReceiverCB.bindenv(this));
-        }
+                local url = "ssl://" + _connStrParsed.HostName;
 
-        function _openAuthSender() {
-            _senders.auth <- _sessions.auth.opensender(AZURE_CLAIM_BASED_SECURITY_PATH, _amqpAuthSenderStatusHandler.bindenv(this));
-        }
+                _state =  AZURE_IOT_CLIENT_STATE_DISCONNECTED | AZURE_IOT_CLIENT_STATE_CONNECTING;
 
-        function _openEventSession() {
-            _sessions.event <- _connection.opensession(_amqpEventSessionStatusHandler.bindenv(this));
+                _mqttclient.connect(url, _connStrParsed.DeviceId, options);
+            } else {
+                try {
+                    // We are connected or connecting now
+                    _onConnectCb(_state & AZURE_IOT_CLIENT_STATE_CONNECTED ? AZURE_IOT_CLIENT_ERROR_ALREADY_CONNECTED : AZURE_IOT_CLIENT_ERROR_OP_NOT_ALLOWED_NOW);
+                } catch (e) {
+                    _error("Exception at onConnect user callback: " + e);
+                }
+            }
         }
 
-        function _openEventSender() {
-            _senders.event <- _sessions.event.opensender(AzureIoTHub.Endpoint.eventPath(_config.deviceId), _amqpEventSenderStatusHandler.bindenv(this));
+        // Closes the connection to Azure IoT Hub. Does nothing if the connection is already closed.
+        //
+        // Returns:                         Nothing.
+        function disconnect() {
+            if (_state != AZURE_IOT_CLIENT_STATE_DISCONNECTED && !(_state & AZURE_IOT_CLIENT_STATE_DISCONNECTING)) {
+                _state = _state | AZURE_IOT_CLIENT_STATE_DISCONNECTING;
+                _mqttclient.disconnect(_onDisconnect.bindenv(this));
+            }
         }
 
-        function _openEventReceiver() {
-            _receivers.event <- _sessions.event.openreceiver(AzureIoTHub.Endpoint.messagePath(_config.deviceId), _amqpEventReceiverStatusHandler.bindenv(this), function(deliveries) {
-                _handleDeliveries(deliveries, _handlers.onEvent);
-            }.bindenv(this));
-        }
+        // Checks if the client is connected to Azure IoT Hub.
+        //
+        // Returns:                         Boolean: true if the client is connected, false otherwise.
+        function isConnected() {
+            return _state & AZURE_IOT_CLIENT_STATE_CONNECTED > 0;
```

Comment:
this is not going to work and will fail with a runtime exception as the `>` operator precede `&`

Reply:
Done